File: Ayeon/바이러스파이프/바이러스파이프.py
from collections import deque

def solution(n, infection, edges, k):
    # 1. 트리 그래프 구축 (adj[노드] = [(연결된 노드, 파이프 타입), ...])
    adj = [[] for _ in range(n + 1)]
    for x, y, t in edges:
        adj[x].append((y, t))
        adj[y].append((x, t))
    
    # 중복 계산을 막기 위한 메모이제이션(DP) 배열
    # memo[node] = 이 노드가 감염되었을 때의 '최소 행동 횟수(turn)'
    # 더 적은 턴 만에 감염될수록 앞으로 더 많은 파이프를 타고 퍼질 수 있으므로 이득
    memo = {}

    # 모든 파이프 선택 조합을 탐색하는 DFS 함수
    # turn: 현재 몇 번째 파이프를 열 차례인지 (0부터 시작)
    # current_infected: 현재까지 감염된 노드들의 집합 {노드: 감염된 turn}
    max_infected_count = 0

    def dfs(turn, current_infected):
        nonlocal max_infected_count
        
        # 현재까지 감염된 개수로 최댓값 갱신
        max_infected_count = max(max_infected_count, len(current_infected))
        
        # 최대 행동 횟수 k번에 도달했다면 종료
        if turn == k:
            return

        # 다음 턴에 열 수 있는 파이프 타입은 1, 2, 3 중 하나
        for pipe_type in [1, 2, 3]:
            # 이번 파이프(pipe_type)를 열었을 때 새로 감염되는 노드들을 찾기 위한 BFS
            queue = deque()
            next_infected = dict(current_infected) # 기존 감염 상태 복사
            
            # 현재 감염된 모든 노드들을 BFS의 시작점으로 넣음
            for node in current_infected:
                queue.append(node)
                
            while queue:
                curr = queue.popleft()
                
                for nxt, t in adj[curr]:
                    # 파이프 타입이 일치하고, 아직 감염되지 않았거나 
                    # 혹은 더 이른 턴(turn)에 감염시킬 수 있는 경우라면 확장
                    if t == pipe_type:
                        if nxt not in next_infected or next_infected[nxt] > turn + 1:
                            next_infected[nxt] = turn + 1
                            queue.append(nxt)
            
            # 가지치기(Optimization): 이전에 완전히 똑같은 감염 상태와 턴을 방문한 적이 있다면 패스
            # 이 문제는 n과 k가 작아서 이 정도로도 충분히 통과
            dfs(turn + 1, next_infected)

    # 초기 상태: 0번째 턴에 infection 노드만 감염됨
    initial_status = {infection: 0}
    dfs(0, initial_status)
    
    return max_infected_count

File: Hoseong/카카오1차/바이러스파이프/바이러스파이프.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <queue>

using namespace std;

struct Edge {
    int to;
    int type;
};

int max_infected = 0;
int N, K;
vector<vector<Edge>> adj;

// [BFS] 현재 감염된 상태에서 특정 타입의 파이프를 열었을 때의 감염 전파를 시뮬레이션
vector<bool> spread(const vector<bool>& current_infected, int type) {
    vector<bool> next_infected = current_infected;
    queue<int> q;
    
    // 현재 감염되어 있는 모든 노드를 큐의 시작점으로 추가
    for (int i = 1; i <= N; ++i) {
        if (current_infected[i]) {
            q.push(i);
        }
    }
    
    // 연결된 파이프가 매개변수로 주어진 type과 일치할 때만 큐에 넣으며 전파
    while (!q.empty()) {
        int curr = q.front();
        q.pop();
        
        for (const auto& edge : adj[curr]) {
            if (edge.type == type && !next_infected[edge.to]) {
                next_infected[edge.to] = true;
                q.push(edge.to);
            }
        }
    }
    
    return next_infected;
}

// [DFS] 백트래킹을 이용한 조합 탐색
void dfs(int step, const vector<bool>& current_infected, int current_count) {
    max_infected = max(max_infected, current_count);
    
    // 조기 종료 조건: 모든 노드가 감염되었거나, 최대 행동 수 K에 도달했을 때
    if (current_count == N || step == K) {
        return;
    }
    
    // 1(A), 2(B), 3(C) 종류의 파이프를 각각 열어보기
    for (int type = 1; type <= 3; ++type) {
        vector<bool> next_infected = spread(current_infected, type);
        
        // 새로 감염된 노드의 개수 세기
        int next_count = 0;
        for (int i = 1; i <= N; ++i) {
            if (next_infected[i]) next_count++;
        }
        
        // 가지치기: 실제로 새로 감염된 배양체가 존재하는 경우에만 깊이 탐색 진행
        if (next_count > current_count) {
            dfs(step + 1, next_infected, next_count);
        }
    }
}

int solution(int n, int infection, vector<vector<int>> edges, int k) {
    // 전역 변수 초기화
    N = n;
    K = k;
    max_infected = 0;
    adj.assign(n + 1, vector<Edge>());
    
    // 인접 리스트 생성 (양방향 그래프)
    for (const auto& edge : edges) {
        int u = edge[0];
        int v = edge[1];
        int type = edge[2];
        adj[u].push_back({v, type});
        adj[v].push_back({u, type});
    }
    
    // 최초 감염 상태 초기화
    vector<bool> start_infected(n + 1, false);
    start_infected[infection] = true;
    
    // DFS 백트래킹 시작 (0단계, 최초 상태, 감염 수 1)
    dfs(0, start_infected, 1);
    
    return max_infected;
}